File: graph/java/AdjacencyListGraph.java
/**
 * @author Ayomide Oyekanmi
 */
class AdjacencyListGraph<E> {
    Map<E, List<E>> graph;

    AdjacencyListGraph(){
        graph = new HashMap<>();
    }

    /**
     * Adds vertex to graph
     * 
     * @param v vertex to add to graph
     */
    public void addVertex(E v){
        if(graph.containsKey(v)){
            return;
        }

        List<E> list = new ArrayList<>();
        graph.put(v, new list);
    }

    /**
     * Returns true if the vertex exists, and false otherwise
     * 
     * @param v The vertex whose presence in the graph we want to ascertain
     * @return boolean true if vertex exists, and false otherwise
     */
    public boolean hasVertex(E v){
        return graph.containsKey(v);
    }

    /**
     * Adds edge to graph
     * 
     * @param v1 The first of the adjacent vertices whose edge we want to add to the graph
     * @param v2 The second of the adjacent vertices whose edge we want to add to the graph
     */
    public void addEdge(E v1, E v2){
        if(!graph.containsKey(v1)){
            List<E> list = new ArrayList<>();
            graph.put(v1, new list);
        }

        if(!graph.containsKey(v1)){
            List<E> list = new ArrayList<>();
            graph.put(v2, new list);
        }

        graph.get(v1).add(v2);
        graph.get(v2).add(v1);
    }

    /**
     * Returns true if the edge exists, and false otherwise
     * 
     * @param v1 The first of the vertices we want to check if an edge exists for
     * @param v2 The second of the vertices we want to check if an edge exists for
     * 
     * @return boolean true if edge exists, and false otherwise
     */
    public boolean hasEdge(E v1, E v2){
        if(!graph.containsKey(v1)){
            return false;
        }
        
        List<E> neighbours = graph.get(v1);
        for(E neighbour : neighbours){
            if(v2.equals(neighbour)){
                return true;
            }
        }

        return false;
    }

    /**
     * Returns all the neighbours (or adjacent nodes) for given vertex
     * 
     * @param v Vertex whose neighbours we want to return
     * @return List<E> a list of neighbours for given vertex
     */
    public List<E> findNeighbours(E v){
        if(!graph.containsKey(v)){
            return new ArrayList<>();
        }

        return graph.get(v);
    }
}

File: graph/C#/AdjacencyListGraph.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Basic_Data_Structure
{
    // @author : Babatunde Ibukun
    class AdjacencyListGraph<T>
    {
        //This is used to represent a graph and every vertices/nodes within it. 
        //This a dictionary that contains a list, each key in the dictionary represent a vertex/node and each 
        //value is used to represent a vertex's adjacent vertices
        public Dictionary<T, List<T>> Graph { get; set; }

        public AdjacencyListGraph()
        {
            this.Graph = new Dictionary<T, List<T>>();
        }


        public int Size
        {
            get { return this.Graph.Count; }
        }

        public void AddVertex(T vertex)
        {
            if (this.Graph.ContainsKey(vertex))
            {
                return;
            }
            List<T> vertices = new List<T>();
            this.Graph.Add(vertex, vertices);
        }

        public bool HasVertex(T vertex)
        {
            return this.Graph.ContainsKey(vertex);            
        }

        public void AddEdge(T vertex, T adjacentVertex)
        {
            if (!this.Graph.ContainsKey(vertex))
            {
                List<T> vertices = new List<T>();
                this.Graph.Add(vertex, vertices);
            }
            if (!this.Graph.ContainsKey(adjacentVertex))
            {
                List<T> vertices = new List<T>();
                this.Graph.Add(adjacentVertex, vertices);
            }
            this.Graph[vertex].Add(adjacentVertex);
            this.Graph[adjacentVertex].Add(vertex);
        }

        public List<T> FindNeighbours(T vertex)
        {
            return this.Graph[vertex];
        }

        public bool HasEdge(T vertex, T vertexToCheck)
        {
            return this.Graph[vertex].Contains(vertexToCheck);
        }

        static void Main1(string[] args)
        {
            AdjacencyListGraph<int> graph = new AdjacencyListGraph<int>();
            graph.AddVertex(3);
            graph.AddVertex(10);
            graph.AddEdge(3, 20);
            graph.AddVertex(20);
            graph.AddEdge(10, 40);
            graph.AddEdge(3, 40);
            graph.AddVertex(40);
            List<int> vertices = graph.FindNeighbours(3); // 20, 40
            Console.WriteLine(graph.HasVertex(20)); // true
            Console.WriteLine(graph.HasEdge(20, 50)); //false
            foreach (int vertex in vertices)
            {
                Console.WriteLine(vertex);
            }
        }
    }
}
